src/12-03.zig

```diff
@@ -0,0 +1,178 @@
+            data: T,
+            next: ?*Node = null,
+            prev: ?*Node = null,
+        };
+
+        const Self = @This();
+
+        head: ?*Node = null,
+        tail: ?*Node = null,
+        len: usize = 0,
+        allocator: Allocator,
+
+        fn init(allocator: Allocator) Self {
+            return .{ .allocator = allocator };
+        }
+
+        fn add(self: *Self, value: T) !void {
+            const node = try self.allocator.create(Node);
+            node.* = Node{
+                .data = value,
+                .prev = self.tail,
+                .next = null,
+            };
+
+            if (self.tail) |*tail| {
+                tail.*.next = node;
+            } else {
+                self.head = node;
+            }
+            self.tail = node;
+
+            self.len += 1;
+        }
+
+        fn remove(self: *Self, value: T) bool {
+            var current = self.head;
+
+            while (current) |cur| : (current = cur.next) {
+                if (cur.data == value) {
+                    // if the current node has a previous node
+                    // then set the previous node's next to the current node's next
+                    if (cur.prev) |prev| {
+                        prev.next = cur.next;
+                    } else {
+                        // if the current node has no previous node
+                        self.head = cur.next;
+                    }
+
+                    // if the current node has a next node
+                    // then set the next node's previous to the current node's previous
+                    if (cur.next) |next| {
+                        next.prev = cur.prev;
+                    } else {
+                        //  if the current node has no next node
+                        // then set the tail to the current node's previous
+                        self.tail = cur.prev;
+                    }
+
+                    self.allocator.destroy(cur);
+                    self.len -= 1;
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        fn len(self: *Self) usize {
+            return self.len;
+        }
+
+        fn search(self: *Self, value: T) bool {
+            var current: ?*Node = self.head;
+
+            while (current) |cur| : (current = cur.next) {
+                if (cur.data == value) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        fn visit(self: *Self, visitor: *const fn (i: usize, v: T) anyerror!void) !void {
+            var current: ?*Node = self.head;
+            var i: usize = 0;
+
+            while (current) |cur| : (i += 1) {
+                try visitor(i, cur.data);
+                current = cur.next;
+            }
+        }
+
+        fn deinit(self: *Self) void {
+            var current: ?*Node = self.head;
+
+            while (current) |cur| : (current = cur.next) {
```

Comment:
When you execute `current = cur.next`, isn't cur already destroyed?

Reply:
yes, I need to set a new variable to store it.